src/gsUtils/gsMesh/gsHalfEdgeMesh.hpp

```diff
@@ -0,0 +1,697 @@
+/** @file gsHalfEdgeMesh.hpp
+
+    @brief Provides implementation of the gsHalfEdgeMesh class.
+
+    This file is part of the G+Smo library.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): L. Groiss, J. Vogl
+*/
+
+#pragma once
+
+namespace gismo
+{
+//**********************************************
+//************ class gsHalfEdgeMesh ************
+//**********************************************
+struct less_than_ptr
+{
+    bool operator()(gsMesh<>::gsVertexHandle lhs, gsMesh<>::gsVertexHandle rhs)
+    {
+        return ((*lhs) < (*rhs));
+    }
+};
+
+struct equal_ptr
+{
+    bool operator()(gsMesh<>::gsVertexHandle lhs, gsMesh<>::gsVertexHandle rhs)
+    {
+        return ((*lhs) == (*rhs));
+    }
+};
+
+//********************************************************************************
+
+template<class T>
+gsHalfEdgeMesh<T>::gsHalfEdgeMesh(const gsMesh<> &mesh, real_t precision)
+    : gsMesh<>(mesh), m_precision(precision)
+{
+    std::sort(this->vertex.begin(), this->vertex.end(), less_than_ptr());
+    typename std::vector<gsVertex<T> *, std::allocator<gsVertex<T> *> >::iterator
+    last = std::unique(this->vertex.begin(), this->vertex.end(), equal_ptr());
+gsDebugVar(this->vertex.size());
+    for (size_t i = 0; i < this->face.size(); i++)
+    {
+        m_halfedges.push_back(getInternHalfedge(this->face[i], 1));
+        m_halfedges.push_back(getInternHalfedge(this->face[i], 2));
+        m_halfedges.push_back(getInternHalfedge(this->face[i], 3));
+    }
+
+    //freeAll(last,this->vertex.end()); // afterwards it fails with invalid read
```

Comment:
Looks like here the memory leak occurs. The vertices from last to vertex.end will not be deletet. Free them here results in invalid read